source/fuzz/fuzzer_pass_add_parameters.cpp

```diff
@@ -69,15 +72,72 @@ void FuzzerPassAddParameters::Apply() {
     auto num_new_parameters =
         GetFuzzerContext()->GetRandomNumberOfNewParameters(
             GetNumberOfParameters(function));
+
     for (uint32_t i = 0; i < num_new_parameters; ++i) {
+      auto current_type_id =
+          type_candidates[GetFuzzerContext()->RandomIndex(type_candidates)];
+      auto current_type =
+          GetIRContext()->get_type_mgr()->GetType(current_type_id);
+      std::map<uint32_t, uint32_t> call_parameter_id;
+
+      // Consider the case when a pointer type was selected.
+      if (current_type->kind() == opt::analysis::Type::kPointer) {
+        auto storage_class = fuzzerutil::GetStorageClassFromPointerType(
+            GetIRContext(), current_type_id);
+        switch (storage_class) {
+          case SpvStorageClassFunction: {
+            // In every caller find or create a local variable with that has the
+            // selected type.
+            for (auto* instr :
+                 fuzzerutil::GetCallers(GetIRContext(), function.result_id())) {
+              auto block = GetIRContext()->get_instr_block(instr);
+              auto function_id = block->GetParent()->result_id();
+              uint32_t variable_id =
+                  FindOrCreateLocalVariable(current_type_id, function_id, true);
+              call_parameter_id[instr->result_id()] = variable_id;
+            }
+          } break;
+          case SpvStorageClassPrivate:
+          case SpvStorageClassWorkgroup: {
+            // If there exists at least one caller, find or create a global
+            // variable that has the selected type.
+            std::vector<opt::Instruction*> callers =
+                fuzzerutil::GetCallers(GetIRContext(), function.result_id());
+            if (!callers.empty()) {
+              uint32_t variable_id =
+                  FindOrCreateGlobalVariable(current_type_id, true);
+              for (auto* instr : callers) {
+                call_parameter_id[instr->result_id()] = variable_id;
+              }
+            }
+          } break;
+          default:
+            break;
+        }
+      } else {
+        // If there exists at least one caller, find or create a zero constant
+        // with that has the selected type.
+        std::vector<opt::Instruction*> callers =
+            fuzzerutil::GetCallers(GetIRContext(), function.result_id());
+        if (!callers.empty()) {
+          uint32_t constant_id =
+              FindOrCreateZeroConstant(current_type_id, true);
+          for (auto* instr :
+               fuzzerutil::GetCallers(GetIRContext(), function.result_id())) {
+            call_parameter_id[instr->result_id()] = constant_id;
+          }
+        }
+      }
+
+      // If the function has no callers, and a selected type to the key 0. It is
+      // necessary to pass information of the new type to the transformation.
+      // Since there are no callers, we pass the type id, not the value id.
```

Comment:
Thanks for spotting this issue. I'm not sure about the current approach to resolving it, though. Maybe it is better to have a separate field with the parameter's type id. I haven't used the type id initially because it was always possible to compute it. In this case, however, using an explicit type id will make the code cleaner. Besides, we still need to check that all values in the `call_parameter_id` have the same type even without an explicit type id field.

Reply:
@antnkarp, @paulthomson, what do you think?